radix-substate-store-interface/src/interface.rs

```diff
@@ -233,14 +267,304 @@ pub trait SubstateDatabase {
     }
 }
 
+impl<T: SubstateDatabase + ?Sized> SubstateDatabaseExtensions for T {}
+
+/// These are a separate trait so that [`SubstateDatabase`] stays object-safe,
+/// and can be used as `dyn SubstateDatabase`.
+///
+/// Generic parameters aren't permitted on object-safe traits.
+pub trait SubstateDatabaseExtensions: SubstateDatabase {
+    fn db_partition_key(
+        node_id: impl AsRef<NodeId>,
+        partition_number: PartitionNumber,
+    ) -> DbPartitionKey {
+        SpreadPrefixKeyMapper::to_db_partition_key(node_id.as_ref(), partition_number)
+    }
+
+    fn db_sort_key<'a>(substate_key: impl ResolvableSubstateKey<'a>) -> DbSortKey {
+        SpreadPrefixKeyMapper::to_db_sort_key_from_ref(
+            substate_key.into_substate_key_or_ref().as_ref(),
+        )
+    }
+
+    fn optional_db_sort_key<'a>(
+        optional_substate_key: impl ResolvableOptionalSubstateKey<'a>,
+    ) -> Option<DbSortKey> {
+        optional_substate_key
+            .into_optional_substate_key_or_ref()
+            .map(|key_or_ref| SpreadPrefixKeyMapper::to_db_sort_key_from_ref(key_or_ref.as_ref()))
+    }
+
+    /// Reads the substate using the default [`SpreadPrefixKeyMapper`].
+    ///
+    /// ## Example use:
+    /// ```ignore
+    /// let is_bootstrapped = store.read_substate(
+    ///     PACKAGE_PACKAGE,
+    ///     TYPE_INFO_FIELD_PARTITION,
+    ///     TypeInfoField::TypeInfo,
+    /// ).is_some();
+    /// ```
+    fn read_substate<'a>(
+        &self,
+        node_id: impl AsRef<NodeId>,
+        partition_number: PartitionNumber,
+        substate_key: impl ResolvableSubstateKey<'a>,
+    ) -> Option<Vec<u8>> {
+        self.get_substate(
+            &Self::db_partition_key(node_id, partition_number),
+            &Self::db_sort_key(substate_key),
+        )
+    }
+
+    /// Reads the substate using the default [`SpreadPrefixKeyMapper`], and then
+    /// decodes the substate value. It panics if the value has a decode error.
+    ///
+    /// ## Example use:
+    /// ```ignore
+    /// let type_info_substate: TypeInfoSubstate = store.read_substate_typed(
+    ///     PACKAGE_PACKAGE,
+    ///     TYPE_INFO_FIELD_PARTITION,
+    ///     TypeInfoField::TypeInfo,
+    /// );
+    /// ```
+    fn read_substate_typed<'a, T: ScryptoDecode>(
+        &self,
+        node_id: impl AsRef<NodeId>,
+        partition_number: PartitionNumber,
+        substate_key: impl ResolvableSubstateKey<'a>,
+    ) -> Option<T> {
+        let raw = self.read_substate(node_id, partition_number, substate_key)?;
+        Some(decode_value(&raw))
+    }
+
+    /// Use `from_substate_key_inclusive: None::<SubstateKey>` to read from start.
+    #[inline]
+    fn read_entries_unknown_key<'a>(
```

Comment:
The `_unknown_key` sounds a bit weird.
Maybe align it with some other naming here? `_untyped_key`? or `_raw_key`?